slave/slave.c

```diff
@@ -0,0 +1,232 @@
+    if (res == NULL) {
+        printf("Query to %s failed\n", args->args->machine_addr);
+        return (void *) 1;
+    }
+    else {
+		printf("Obtained result: %d\n", res->exit_code);
+        results[args->query_result_index] = res;
+    }
+    return (void *) 0;
+}
+
+query_result *rq_range_root_1_svc(rq_range_root_args query, struct svc_req *req)
+{
+    int num_threads = query.num_ranges;
+    unsigned int *range_array = query.range_array.range_array_val;
+    pthread_t tids[num_threads];
+    // the results to be anded together (in general, OP)
+    results = (query_result **) malloc(sizeof(query_result *) * num_threads);
+    int i;
+    int array_index = 0;
+    for (i = 0; i < num_threads; i++) {
+        int num_nodes = range_array[array_index++];
+        rq_pipe_args *pipe_args = (rq_pipe_args *) malloc(sizeof(rq_pipe_args) * num_nodes);
+        rq_pipe_args *head_args = pipe_args;
+        int j;
+        for (j = 0; j < num_nodes; j++) {
+            pipe_args->machine_addr = SLAVE_ADDR[range_array[array_index++]];
+            pipe_args->vec_id = range_array[array_index++];
+            pipe_args->op = '|';
+            if (j < num_nodes - 1) {
+                rq_pipe_args *next_args = (rq_pipe_args *) malloc(sizeof(rq_pipe_args) * num_nodes);
+                pipe_args->next = next_args;
+                pipe_args = next_args;
+            }
+            else {
+                pipe_args->next = NULL;
+            }
+        }
+
+        coord_thread_args *thread_args = (coord_thread_args *) malloc(sizeof(coord_thread_args));
+        thread_args->query_result_index = i;
+        thread_args->args = head_args;
+
+        // allocate the appropriate number of args
+        pthread_create(&tids[i], NULL, init_coordinator_thread, (void *) thread_args);
+    }
+    for (i = 0; i < num_threads; i++) pthread_join(tids[i], NULL);
+    // TODO AND all the results together
+    //query_result *res = (query_result *) malloc(sizeof(query_result));
+    return results[0];
+}
+
+#define TESTING_SLOW_PROC 0
+
+int result;
+
+int *commit_msg_1_svc(int message, struct svc_req *req)
+{
+	printf("SLAVE: VOTING\n");
+    int ready = 1; // test value
+    if (ready) {
+        result = VOTE_COMMIT;
+    }
+    else {
+        /* possible reasons: lack of memory, ... */
+        result = VOTE_ABORT;
+    }
+    /* make the process run slow, to test failure */
+    if (TESTING_SLOW_PROC) {
+        sleep(TIME_TO_VOTE + 1);
+    }
+    return &result;
+
+}
+
+int *commit_vec_1_svc(struct commit_vec_args args, struct svc_req *req)
+{
+	printf("SLAVE: Putting vector %u\n", args.vec_id);
+    FILE *fp;
+    char filename_buf[128];
+    snprintf(filename_buf, 128, "v_%d.dat", args.vec_id); // XXX: function to get vector filename
+    fp = fopen(filename_buf, "wb");
+    char buffer[128];
+    int i;
+    for (i = 0; i < args.vector.vector_len; i++)
+        snprintf(buffer, 128, "%llu", args.vector.vector_val[i]);
```

Comment:
We are getting the following error:
```
slave/slave.c:226:31: warning: format ‘%llu’ expects argument of type ‘long long unsigned int’, but argument 4 has type ‘u_quad_t {aka long unsigned int}’
```
Not an issue; but we could format it as `%lu` if we don't want to see the warning.